cities/modeling/modeling_utils.py

```diff
@@ -0,0 +1,199 @@
+        i_o_data.drop(columns=columns_to_drop.columns, inplace=True)
+
+    i_o_data.rename(columns={"Value": outcome_dataset}, inplace=True)
+
+    i_o_data["state"] = [code // 1000 for code in i_o_data["GeoFIPS"]]
+
+    N_s = len(i_o_data["state"].unique())  # number of states
+    i_o_data.dropna(inplace=True)
+
+    i_o_data["unit_index"] = pd.factorize(i_o_data["GeoFIPS"].values)[0]
+    i_o_data["state_index"] = pd.factorize(i_o_data["state"].values)[0]
+    i_o_data["time_index"] = pd.factorize(i_o_data["Year"].values)[0]
+
+    assert i_o_data["GeoFIPS"].isin(f_covariates_joint["GeoFIPS"]).all()
+
+    f_covariates_joint.drop(columns=["GeoName"], inplace=True)
+    data = i_o_data.merge(f_covariates_joint, on="GeoFIPS", how="left")
+
+    assert not data.isna().any().any()
+
+    time_index_idx = data.columns.get_loc("time_index")
+    covariates_df = data.iloc[:, time_index_idx + 1 :].copy()
+    covariates_df_sparse = covariates_df.copy()
+    covariates_df_sparse["unit_index"] = data["unit_index"]
+    covariates_df_sparse["state_index"] = data["state_index"]
+    covariates_df_sparse.drop_duplicates(inplace=True)
+    assert set(covariates_df_sparse["unit_index"]) == set(data["unit_index"])
+
+    # get tensors
+
+    if torch.cuda.is_available():
+        device = torch.device("cuda")
+    else:
+        device = torch.device("cpu")
+
+    y = data[f"{outcome_dataset}_shifted_by_{forward_shift}"]
+    y = torch.tensor(y, dtype=torch.float32, device=device)
+
+    unit_index = torch.tensor(data["unit_index"], dtype=torch.int, device=device)
+    unit_index_sparse = torch.tensor(
+        covariates_df_sparse["unit_index"], dtype=torch.int
+    )
+
+    state_index = torch.tensor(data["state_index"], dtype=torch.int, device=device)
+    state_index_sparse = torch.tensor(
+        covariates_df_sparse["state_index"], dtype=torch.int
+    )
+
+    time_index = torch.tensor(data["time_index"], dtype=torch.int, device=device)
+    intervention = torch.tensor(
+        data[intervention_variable], dtype=torch.float32, device=device
+    )
+
+    covariates = torch.tensor(covariates_df.values, dtype=torch.float32, device=device)
+
+    covariates_df_sparse.drop(columns=["unit_index", "state_index"], inplace=True)
+    covariates_sparse = torch.tensor(
+        covariates_df_sparse.values, dtype=torch.float32, device=device
+    )
+
+    N_cov = covariates.shape[1]  # number of covariates
+    N_u = covariates_sparse.shape[0]  # number of units (counties)
+    N_obs = len(y)  # number of observations
+    N_t = len(time_index.unique())  # number of time points
+    N_s = len(state_index.unique())  # number of states
+
+    assert len(intervention) == len(y)
+    assert len(unit_index) == len(y)
+    assert len(state_index) == len(unit_index)
+    assert len(time_index) == len(unit_index)
+    assert covariates.shape[1] == covariates_sparse.shape[1]
+    assert len(unit_index_sparse) == N_u
+
+    return {
```

Comment:
I don't love returning giant dicts but also don't have a better idea. Leave for now probably.